File: lib/java/src/ae/math/StaticBackend.java
package ae.math;

public final class StaticBackend extends VectorBackend {
	
	private final float[] _data = new float[4];

	@Override
	protected final float _getElement(final int index) {return _data[index];}
	
	@Override
	protected final float _getX() {return _data[0];}
	@Override
	protected final float _getY() {return _data[1];}
	@Override
	protected final float _getZ() {return _data[2];}
	@Override
	protected final float _getW() {return _data[3];}
	
	@Override
	protected final void _setElement(
			final int   index,
			final float value) {
		
		_data[index] = value;
	}
	
	@Override
	protected final void _setX(final float x) {_data[0] = x;}
	@Override
	protected final void _setY(final float y) {_data[1] = y;}
	@Override
	protected final void _setZ(final float z) {_data[2] = z;}
	@Override
	protected final void _setW(final float w) {_data[3] = w;}

	public StaticBackend(
			final VectorBackend backend) {
		
		this(backend.getX(), backend.getY(), backend.getZ(), backend.getW());
	}
	
	public StaticBackend(
			final float x,
			final float y,
			final float z,
			final float w) {
		
		_data[0] = x;
		_data[1] = y;
		_data[2] = z;
		_data[3] = w;
	}
}

File: lib/javascript/src/math/VectorBackend/StaticBackend.js

ae.math.StaticBackend = class StaticBackend extends ae.math.VectorBackend {
	
	_data: Array<number>;
	
	constructor(
			xOrBackend: (number|ae.math.VectorBackend),
			y:          number = 0,
			z:          number = 0,
			w:          number = 0) {
		
		super();
		
		this._data = xOrBackend instanceof ae.math.VectorBackend ?
			[xOrBackend.x, xOrBackend.y, xOrBackend.z, xOrBackend.w] :
			[xOrBackend,   y,            z,            w];
	}
	
	get x(): number {return this._data[0];}
	get y(): number {return this._data[1];}
	get z(): number {return this._data[2];}
	get w(): number {return this._data[3];}
	
	set x(x: number): ae.math.VectorBackend {
		
		this._data[0] = x;
		return this;
	}
	
	set y(y: number): ae.math.VectorBackend {
		
		this._data[1] = y;
		return this;
	}
	
	set z(z: number): ae.math.VectorBackend {
		
		this._data[2] = z;
		return this;
	}
	
	set w(w: number): ae.math.VectorBackend {
		
		this._data[3] = w;
		return this;
	}
	
	getElement(index: number): number {
		
		return this._data[index];
	}
	
	setElement(
			index: number,
			value: number): ae.math.VectorBackend {
		
		this._data[index] = value;
		
		return this;
	}
};
